sockeye/decoder.py

```diff
@@ -249,62 +236,98 @@ def _decode(self,
+        :param step: Global step of inference procedure.
         :param target_embed_prev: Previous target word embedding. Shape: (batch_size, target_num_embed).
         :param source_encoded_max_length: Length of encoded source time dimension.
         :param states: Arbitrary list of decoder states.
         :return: logit inputs, attention probabilities, next decoder states.
         """
-        source_encoded, source_encoded_lengths = states
+        source_encoded, source_encoded_lengths = states[:2]  # pylint: disable=unbalanced-tuple-unpacking
 
-        # indices: (batch_size,)
-        indices = target_embed_lengths - 1  # type: mx.sym.Symbol
-
-        # (batch_size, target_max_length, 1)
-        mask = mx.sym.expand_dims(mx.sym.one_hot(indices=indices,
-                                                 depth=target_embed_max_length,
-                                                 on_value=1, off_value=0), axis=2)
+        symbolic_step = mx.sym.arange(start=step - 1, stop=step, step=1, name='symbolic_step')
+        # (batch_size, num_embed)
+        target_embed_prev = self.pos_embedding.encode_positions(symbolic_step, target_embed_prev)
+        # (batch_size, 1, num_embed)
+        target = mx.sym.expand_dims(target_embed_prev, axis=1)
+
+        # (batch_size * heads, max_length)
+        source_bias = transformer.get_variable_length_bias(lengths=source_encoded_lengths,
+                                                           max_length=source_encoded_max_length,
+                                                           num_heads=self.config.attention_heads,
+                                                           fold_heads=True,
+                                                           name="%ssource_bias" % self.prefix)
+        # (batch_size * heads, 1, max_length)
+        source_bias = mx.sym.expand_dims(source_bias, axis=1)
+
+        # auto-regressive bias for last position in sequence
+        # (1, target_max_length, target_max_length)
+        target_bias = transformer.get_autoregressive_bias(step, name="%sbias" % self.prefix)
+        target_bias = mx.sym.slice_axis(target_bias, axis=1, begin=-1, end=step)
 
-        # (batch_size, target_max_length, model_size)
-        target = self._decode(source_encoded, source_encoded_lengths, source_encoded_max_length,
-                              target_embed, target_embed_max_length)
+        layer_caches = self._get_layer_caches_from_states(list(states))
+        cache = []  # type: List[mx.sym.Symbol]
+        for layer, layer_cache in zip(self.layers, layer_caches):
+            target = layer(target=target,
+                           target_bias=target_bias,
+                           source=source_encoded,
+                           source_bias=source_bias,
+                           cache=layer_cache)
+            cache += [layer_cache['k'], layer_cache['v']]
+        cache = mx.sym.concat(*cache, dim=1, name='new_cache')
```

Comment:
A note reminding us what we're adding to the cache (conceptually) would be helpful too.

Reply:
done